include/deal.II/lac/trilinos_tpetra_sparsity_pattern.h

```diff
@@ -0,0 +1,1378 @@
+
+      /**
+       * Return the number of nonzero elements of this sparsity pattern.
+       */
+      std::uint64_t
+      n_nonzero_elements() const;
+
+      /**
+       * Return the number of entries in the given row.
+       *
+       * In a parallel context, the row in question may of course not be
+       * stored on the current processor, and in that case it is not
+       * possible to query the number of entries in it. In that case,
+       * the returned value is `static_cast<size_type>(-1)`.
+       */
+      size_type
+      row_length(const size_type row) const;
+
+      /**
+       * Compute the bandwidth of the matrix represented by this structure. The
+       * bandwidth is the maximum of $|i-j|$ for which the index pair $(i,j)$
+       * represents a nonzero entry of the matrix. Consequently, the maximum
+       * bandwidth a $n\times m$ matrix can have is $\max\{n-1,m-1\}$.
+       */
+      size_type
+      bandwidth() const;
+
+      /**
+       * Return whether the object is empty. It is empty if no memory is
+       * allocated, which is the same as when both dimensions are zero.
+       */
+      bool
+      empty() const;
+
+      /**
+       * Return whether the index (<i>i,j</i>) exists in the sparsity pattern
+       * (i.e., it may be nonzero) or not.
+       */
+      bool
+      exists(const size_type i, const size_type j) const;
+
+      /**
+       * Return whether a given @p row is stored in the current object
+       * on this process.
+       */
+      bool
+      row_is_stored_locally(const size_type i) const;
+
+      /**
+       * Determine an estimate for the memory consumption (in bytes) of this
+       * object. Currently not implemented for this class.
+       */
+      std::size_t
+      memory_consumption() const;
+
+      /** @} */
+      /**
+       * @name Adding entries
+       */
+      /** @{ */
+      /**
+       * Add the element (<i>i,j</i>) to the sparsity pattern.
+       */
+      void
+      add(const size_type i, const size_type j);
+
+
+      /**
+       * Add several elements in one row to the sparsity pattern.
+       */
+      template <typename ForwardIterator>
+      void
+      add_entries(const size_type row,
+                  ForwardIterator begin,
+                  ForwardIterator end,
+                  const bool      indices_are_sorted = false);
+
+      virtual void
+      add_row_entries(
+        const dealii::types::global_dof_index                  &row,
+        const ArrayView<const dealii::types::global_dof_index> &columns,
+        const bool indices_are_sorted = false) override;
+
+      using SparsityPatternBase::add_entries;
+
+      /** @} */
+      /**
+       * @name Access of underlying Trilinos data
+       */
+      /** @{ */
+
+      /**
+       * Return a Teuchos::RCP to the underlying Trilinos Tpetra::CrsGraph
+       * data that stores the sparsity pattern.
+       */
+      Teuchos::RCP<GraphType>
+      trilinos_sparsity_pattern() const;
```

Comment:
I thought of it more as a question of what pattern we want to use in general. For all non-scalar types, getter functions that are `const` should return a `const` reference. It may not matter in this specific case, but I think as a general rule it is the right thing to do and it would be nice if we did the same here :-)

Reply:
We could return a `Teuchos::RCP<const ...>` in those places.